### DISCUSSION TITLE
How to get event loop with current context

### DISCUSSION THREAD

**Discussion by ed2050** [zauberzeug/nicegui#4117] [Q&A]
### Question

I've been beating my head against the wall all night.  How do you get the asyncio event loop with current nicegui context?  I keep getting same error:

```
Traceback (most recent call last):
  File "/Users/ed2050/gui/folders.py", line 325, in get_dlid
    dlid = await ui.run_javascript ('get_dlid () ;')
  File "/Users/ed2050/Library/Python/3.8/lib/python/site-packages/nicegui/functions/javascript.py", line 22, in run_javascript
    client = globals.get_client()
  File "/Users/ed2050/Library/Python/3.8/lib/python/site-packages/nicegui/globals.py", line 104, in get_client
    return get_slot().parent.client
  File "/Users/ed2050/Library/Python/3.8/lib/python/site-packages/nicegui/globals.py", line 100, in get_slot
    return get_slot_stack()[-1]
IndexError: list index out of range
```

From other discussions [like this](https://github.com/zauberzeug/nicegui/discussions/302), it seems the problem is missing context.  I've tried everything I can think of to get current nicegui event loop with current page context.  I searched nicegui code, boards, docs, etc.  No luck.

**Comment by ed2050** [comment]
More background - this is the code I'm trying to run.

```py
async def get_dlid () :
    'retrieve data from client'
    dlid = await ui.run_javascript ('get_dlid () ;')  # retrieve data from client side script
    return dlid

def runit (func) :
    if inspect.iscoroutinefunction (func) :
        loop = asyncio.get_running_loop ()
        #loop = nicegui.core.loop ()    # new way ?
        #loop = nicegui.globals.loop   # old way ?  doesnt matter, same context error
    
        ret  = loop.run_until_complete (func ())
    else :
        ret = func ()
    
    return ret
```

Why am I doing this?  Because I pass function args that can be sync or async.  I need to call either type from one place, and this is how to do it.  (Also, I'm tired of async crap propagating changes throughout my entire code base).

With other async code (aka not nicegui), this approach works great.  There's no magic context needed.

**Reply by falkoschindler** [reply]
How is `runit` called?
The conceptual problem is: If there is no open UI context when `runit` is called (which would be the case if, e.g., a click event handler would call `runit`), how should `get_dlid` (I assume this is passed as `func` to `runit`) know which client to talk to?

**Reply by ed2050** [reply]
Correct, runit is called from a click handler.  However some context is available.  I save `client` and `request` objects in page state variable `ps`.  This is available when `runit` is called, though not currently passed to it.

Simplified example:
```py
class Pagestate : pass

@ui.page ('/foo')
def foopage (request : Request, client : nicegui.Client) :

    ps = Pagestate ()
    ps.request = request
    ps.client = client

    ui.button ('show dlid', on_click = lambda : ui.notify (runit (get_dlid) ))
```

Note that in practice, `runit` call is actually nested several levels deep in the callstack.  In above example, you could access `client` and `request` directly inside lambda.  In real code, lambda calls `bar (ps)` which calls `rab (ps)` ... until finally something calls `runit`.

Next example is closer to real structure (still greatly simplified):
```py
class Pagestate : pass

def bar (ps) :
    func = ps.getdata
    data = runit (func)
    return data

@ui.page ('/foo')
def foopage (request : Request, client : nicegui.Client) :

    ps = Pagestate ()
    ps.request = request
    ps.client = client
    ps.getdata = get_dlid

    ui.button ('show dlid', on_click = lambda : ui.notify (bar (ps)))
```

On pages like this example, getdata calls a client-side js func.  On other pages, getdata retrieves data locally (server side).  Depends on the page.  Could be async, could be sync.

In the case of `get_dlid`, I suppose I could pass `ps` to runit, runit could pass `ps` to `func`, and `func / get_dlid` could call `ps.client.run_javascript` instead of `ui.run_javascript`.  I'd have to change every other func assigned to `getdata` to also accept `ps` param.  It's doable.  

But that only fixes cases where func calls `run_javascript`.  I'd like a more general solution to cover cases where `func` is server-side async, ie not `run_javascript`.  Would `asyncio.get_running_loop ()` work for server-side cases?  Or would it still be missing context?

**Reply by falkoschindler** [reply]
Ok, so `run_it` has the right context, but `func` might not. But when testing the following example, I see two problems:
```py
def a():
    time.sleep(0.1)
    return 'A'

async def b():
    await asyncio.sleep(0.1)
    return 'B'

def run_it(f: Callable):
    if asyncio.iscoroutinefunction(f):
        return core.loop.run_until_complete(f())
    else:
        return f()

@ui.page('/')
def page():
    ui.button('Run A!', on_click=lambda: ui.notify(run_it(a)))
    ui.button('Run B!', on_click=lambda: ui.notify(run_it(b)))
```

1. Asyncio complains that the loop is already running. So you don't seem to be able to call `run_until_complete`.
2. You're trying to call a potentially async function `func` from a synchronous function `run_it`. This would be ok if `func` is sent to a background task and never returns. But if you want to get a result, you need to pause the main thread, which is usually not desired. And while the main thread is paused, `func` won't be able to run.

It is definitely easier to work the other way around:
```py
def a():
    time.sleep(0.1)
    return 'A'

async def b():
    await asyncio.sleep(0.1)
    return 'B'

async def run_it(f: Callable):
    if asyncio.iscoroutinefunction(f):
        result = await f()
    else:
        result = await run.cpu_bound(f)
    ui.notify(result)

@ui.page('/')
def page():
    ui.button('Run A!', on_click=lambda: run_it(a))
    ui.button('Run B!', on_click=lambda: run_it(b))
```

Now `run_it` is async and the click handler will automatically run it on a background thread. The result of an sync `f` is simply awaited, while the result of a synchronous `f` is awaited using `run.cpu_bound`. Because there's noone waiting for a result from `run_it`, I moved the notification inside this function. Alternatively you could use another wrapper around `run_it`.

**Reply by ed2050** [reply]
> 1. Asyncio complains that the loop is already running. So you don't seem to be able to call `run_until_complete`.

That's why I call `asyncio.get_running_loop ()` instead of using `core.loop`.  It's the same error as calling `asyncio.run ()` after the loop starts.

> 2. You're trying to call a potentially async function `func` from a synchronous function `run_it`. This would be ok if `func` is sent to a background task and never returns. But if you want to get a result, you need to pause the main thread, which is usually not desired. And while the main thread is paused, `func` won't be able to run.

Isn't it the opposite?  `runit` is screwed if `func` never returns - runit will wait forever for a return value.  Then again, same problem happens if synchronous `func` never returns.

But let's say there's some potential for deadlock in passing control from synchronous main thread to async func.  Ok maybe?  I don't know - not a python designer, don't know much about GIL.

What I do know is that `runit ()` works perfectly in another project of mine.  Here's the actual working code, minus irrelevant lines:

```py
async def load_session () :
    await blink.start ()  # interface to blink camera library

def runit (coroutine) :
    'run async coroutine and return'

    loop = asyncio.get_event_loop ()
    ret = loop.run_until_complete (coroutine)
    return ret

def main (args) :
    session = runit (load_session ())
```
 
> It is definitely easier to work the other way around:  _ie calling sync functions from async function_

Depends on your definition of 'easier'. 😄  I considered that.  Unfortunately that propagates `async` to many defs throughout my codebase.  The cure is worse than the disease.

You gave me an idea though.  From problem 2 about blocking main thread.  I wonder if runit can launch a new thread to run the async func and `thread.wait ()` for return value?  Worth a try. 🤷

**Reply by ed2050** [reply]
[Here's a helpful discussion](https://stackoverflow.com/a/71489745/7663296) on calling async func from sync func.  They use same method as runit.

Creating a new thread [doesn't help](https://stackoverflow.com/questions/59645272/how-do-i-pass-an-async-function-to-a-thread-target-in-python).  Can't start a thread with an `async` func as target.  Solution is just to launch a new async task - either with `asyncio.run ()` when no loop started, otherwise `asyncio.get_running_loop ().run_until_complete ()`.  Back to square one.

**Reply by falkoschindler** [reply]
> That's why I call asyncio.get_running_loop () instead of using core.loop.

I don't think there's a difference. The loop is already running, so you can't call `run_until_complete`.

Let's assume this _would_ run without exceptions (maybe using some multi-threading technique):
```py
def async_to_sync(awaitable):
    loop = asyncio.get_running_loop()
    return loop.run_until_complete(awaitable)  # insert magic here

def handle_click():
    print('running...')
    ui.notify('running...')
    async_to_sync(asyncio.sleep(3.0))
    ui.notify('done')
    print('done')

ui.button('Run', on_click=handle_click)
```

Because `handle_click` is not async, nothing else on the event loop can run while `async_to_sync` is being called. The whole UI freezes for three seconds and the first notification will popup only after `handle_click` returns.

**Reply by ed2050** [reply]
> > That's why I call asyncio.get_running_loop () instead of using core.loop.
> 
> I don't think there's a difference. The loop is already running, so you can't call `run_until_complete`.

Don't know what to tell you.  It [works](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_until_complete) for [me](https://stackoverflow.com/a/71489745/7663296).  Code runs perfectly.

From pydocs:
```
loop.run_until_complete(future)
    Run until the future (an instance of Future) has completed.
```

There's also [run_coroutine_threadsafe](https://docs.python.org/3/library/asyncio-task.html#asyncio.run_coroutine_threadsafe).  Tested and it works for submitting a task to a running loop.  I'll try it with nicegui and see what happens. 

> Because `handle_click` is not async, nothing else on the event loop can run while `async_to_sync` is being called. The whole UI freezes for three seconds and the first notification will popup only after `handle_click` returns.

Ok, it may not be a good idea in general.  But if my freeze times are small and I'm willing to pay the price, it can be the lesser of two evils.

What's really the difference between sync funcs and async funcs?  A difference of quality or of degree?  It's just how many times func can be interrupted.  Async 1 or more times, sync 0 times.  But write an async func with no awaits and no I/O calls.  Boom, you have an async func that can be interrupted 0 times.  In other words, it's sync.  There's no qualitative difference.  That's why `async` keyword is an albatross.  Python interpreter should get rid of useless incompatible declarations, treat every func as async, and be done with it.

If you start from a blank slate, ok, make every func async.  Just throw it in all your defs so interpreter is happy.  I don't have that luxury.  I have a large existing codebase of libs my nicegui project needs to use.  I can't just change all of the lib defs and break other apps using them.  So I'm stuck in this two-tier hell where async funcs can call sync but not vice versa.  I'm trying to break out of that prison.  Hope you understand.
